src/kokkos/test/kokkos/VertexFinder_t.cc

```diff
@@ -0,0 +1,299 @@
+                                                  float chi2max,  // max normalized distance to cluster,
+                                                  const member_type& team_member) {
+  clusterTracksByDensity(vdata, vws, minT, eps, errmax, chi2max, team_member);
+  team_member.team_barrier();
+
+  fitVertices(vdata, vws, 50., team_member);
+  team_member.team_barrier();
+
+  splitVertices(vdata, vws, 9.f, team_member);
+  team_member.team_barrier();
+
+  fitVertices(vdata, vws, 5000., team_member);
+  team_member.team_barrier();
+
+  sortByPt2(vdata, vws, team_member);
+}
+#endif
+
+struct Event {
+  std::vector<float> zvert;
+  std::vector<uint16_t> itrack;
+  std::vector<float> ztrack;
+  std::vector<float> eztrack;
+  std::vector<float> pttrack;
+  std::vector<uint16_t> ivert;
+};
+
+struct ClusterGenerator {
+  explicit ClusterGenerator(float nvert, float ntrack)
+      : rgen(-13., 13), errgen(0.005, 0.025), clusGen(nvert), trackGen(ntrack), gauss(0., 1.), ptGen(1.) {}
+
+  void operator()(Event& ev) {
+    int nclus = clusGen(reng);
+    ev.zvert.resize(nclus);
+    ev.itrack.resize(nclus);
+    for (auto& z : ev.zvert) {
+      z = 3.5f * gauss(reng);
+    }
+
+    ev.ztrack.clear();
+    ev.eztrack.clear();
+    ev.ivert.clear();
+    for (int iv = 0; iv < nclus; ++iv) {
+      auto nt = trackGen(reng);
+      ev.itrack[nclus] = nt;
+      for (int it = 0; it < nt; ++it) {
+        auto err = errgen(reng);  // reality is not flat....
+        ev.ztrack.push_back(ev.zvert[iv] + err * gauss(reng));
+        ev.eztrack.push_back(err * err);
+        ev.ivert.push_back(iv);
+        ev.pttrack.push_back((iv == 5 ? 1.f : 0.5f) + ptGen(reng));
+        ev.pttrack.back() *= ev.pttrack.back();
+      }
+    }
+    // add noise
+    auto nt = 2 * trackGen(reng);
+    for (int it = 0; it < nt; ++it) {
+      auto err = 0.03f;
+      ev.ztrack.push_back(rgen(reng));
+      ev.eztrack.push_back(err * err);
+      ev.ivert.push_back(9999);
+      ev.pttrack.push_back(0.5f + ptGen(reng));
+      ev.pttrack.back() *= ev.pttrack.back();
+    }
+  }
+
+  std::mt19937 reng;
+  std::uniform_real_distribution<float> rgen;
+  std::uniform_real_distribution<float> errgen;
+  std::poisson_distribution<int> clusGen;
+  std::poisson_distribution<int> trackGen;
+  std::normal_distribution<float> gauss;
+  std::exponential_distribution<float> ptGen;
+};
+
+// a macro SORRY
+#define LOC_WS(M) ((char*)(ws_h.data()) + offsetof(WorkSpace, M))
+
+void test() {
+  Kokkos::View<ZVertices*, KokkosExecSpace> onGPU_d("onGPU_d", 1);
+  Kokkos::View<WorkSpace*, KokkosExecSpace> ws_d("ws_d", 1);
+
+  Event ev;
+
+  float eps = 0.1f;
+  std::array<float, 3> par{{eps, 0.01f, 9.0f}};
+  for (int nav = 30; nav < 80; nav += 20) {
+    ClusterGenerator gen(nav, 10);
+
+    for (int i = 8; i < 20; ++i) {
+      auto kk = i / 4;  // M param
+
+      gen(ev);
+
+      Kokkos::parallel_for(
+          "init", team_policy(1, 1), KOKKOS_LAMBDA(const member_type& team_member) {
```

Comment:
Guessing
```suggestion
          "init", team_policy(KokkosExecSpace(), 1, 1), KOKKOS_LAMBDA(const member_type& team_member) {
```

Reply:
`using team_policy = Kokkos::TeamPolicy<KokkosExecSpace>;` So execution space is not needed here

Reply:
The difference is that an execution space object is passed to the policy, and in that way when `fence()` is used internally it is called only on that execution space instead of all execution spaces. With CUDA execution space constructed with CUDA stream the call might even be asynchronous (I'm not fully sure in `parallel_for()` case).